spec/services/fetch_coronavirus_statistics_service_spec.rb

```diff
@@ -0,0 +1,124 @@
+RSpec.describe FetchCoronavirusStatisticsService do
+  describe ".call" do
+    it "returns a Statistics object" do
+      body = {
+        data: [
+          { "date" => "2021-03-18", "cumulativeVaccinations" => nil, "hospitalAdmissions" => nil, "newPositiveTests" => 6303 },
+          { "date" => "2021-03-17", "cumulativeVaccinations" => 25_735_472, "hospitalAdmissions" => nil, "newPositiveests" => 5758 },
+          { "date" => "2021-03-16", "cumulativeVaccinations" => 25_273_226, "hospitalAdmissions" => nil, "newPositiveests" => 5294 },
+          { "date" => "2021-03-15", "cumulativeVaccinations" => 24_839_906, "hospitalAdmissions" => nil, "newPositiveests" => 5089 },
+          { "date" => "2021-03-14", "cumulativeVaccinations" => 24_453_221, "hospitalAdmissions" => 426, "newPositiveests" => 4618 },
+          { "date" => "2021-03-13", "cumulativeVaccinations" => 24_196_211, "hospitalAdmissions" => 460, "newPositiveests" => 5534 },
+        ],
+      }
+
+      stub_request(:get, /coronavirus.data.gov.uk/)
+        .to_return(status: 200, body: body.to_json)
+
+      statistics = described_class.call
+      expect(statistics).to be_a(described_class::Statistics)
+      expect(statistics).to have_attributes(
+        cumulative_vaccinations: 25_735_472,
+        cumulative_vaccinations_date: Date.new(2021, 3, 17),
+        hospital_admissions: 426,
+        hospital_admissions_date: Date.new(2021, 3, 14),
+        new_positive_tests: 6303,
+        new_positive_tests_date: Date.new(2021, 3, 18),
+      )
+    end
+
+    context "when there is missing data" do
+      it "returns nil for no data" do
+        stub_request(:get, /coronavirus.data.gov.uk/)
+          .to_return(status: 200, body: { data: [] }.to_json)
+
+        expect(described_class.call).to be_nil
+      end
+
+      it "sets only the fields that have data" do
```

Comment:
It's good that it can deal with this situation but are we ever expecting it to happen? IE can some fields on coronavirus.data.gov.uk return nil whilst others have values? Also, can the frontend deal with this? Will it show headings for all fields or just those that have data?

I'm not sure what the user would think if some values showed whilst others didn't.

Reply:
Yup, the frontend can deal with this - we have conditionals for each option and just won't show a block if there isn't the  data. I don't think a user would be particularly concerned if some showed and not others, we don't tell them what to expect to see.

The API does currently return some fields null as per output actual output from: https://github.com/alphagov/collections/pull/2315/files#diff-583212f27e5e2c931d9ead6be0a3e8b8b55aac9420e3e1dc277d4259004acbb2R5-R13

Question is whether the API could return with all the fields being null, that'd seem very unlikely if it was all operating ok as they'd have to not update the data for a very long time. However given it's an external service and it would be correct to their schema to all be null it feels like it should be handled.